lib/shared/serialization.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * @fileoverview Object serialization utils.
+ * @author Bryan Mishkin
+ */
+
+"use strict";
+
+/**
+ * Check if a value is a primitive or plain object created by the Object constructor.
+ * @param {any} val the value to check
+ * @returns {boolean} true if so
+ * @private
+ */
+function isSerializablePrimitiveOrPlainObject(val) {
+    return (
+        val === null ||
+        typeof val === "string" ||
+         typeof val === "boolean" ||
+         typeof val === "number" ||
+         (typeof val === "object" && val.constructor === Object) ||
+         Array.isArray(val)
+    );
+}
+
+/**
+ * Check if an object is serializable.
+ * Functions or objects like RegExp cannot be serialized by JSON.stringify().
+ * Inspired by: https://stackoverflow.com/questions/30579940/reliable-way-to-check-if-objects-is-serializable-in-javascript
+ * @param {any} obj the object
+ * @returns {boolean} true if the object is serializable
+ */
+function isSerializable(obj) {
+    if (!isSerializablePrimitiveOrPlainObject(obj)) {
+        return false;
+    }
+    for (const property in obj) {
```

Comment:
nits: it's better to use `for...of` to iter arrays - `for...in` is more commonly used for just objects.

Reply:
This is being used to iterate both objects and arrays at this point so I can't simply swap to `for...of` without more refactoring.

Reply:
@aladdin-add please don't nitpick. If something is important, then definitely mention it, but if it's not, it's best to just not mention it.